frame/composable-traits/src/dex.rs

```diff
@@ -111,3 +112,28 @@ pub struct ConstantProductPoolInfo<AccountId, AssetId> {
 	/// Amount of the fee pool charges for the exchange
 	pub fee: Permill,
 }
+
+#[derive(Encode, Decode, TypeInfo, Clone, PartialEq, Eq, RuntimeDebug)]
+pub enum DexRouteNode<PoolId> {
+	Curve(PoolId),
+	Uniswap(PoolId),
+}
+
+/// Describes route for DEX.
+/// `Direct` gives vector of pool_id to use as router.
+#[derive(Encode, Decode, TypeInfo, Clone, PartialEq, Eq, RuntimeDebug)]
+pub enum DexRoute<PoolId, MaxHops: Get<u32>> {
+	Direct(BoundedVec<DexRouteNode<PoolId>, MaxHops>),
+}
+
+pub trait DexRouter<AssetId, PoolId> {
```

Comment:
Should add `buy` and `sell` functions, which route and execute.

Reply:
can we have a single `exchange` method instead of buy/sell ? if not could you please highlight differences?

Reply:
there are 2 desings - have buy+sell vs order(direction).

First can be modelled as trait (see Sell trait) and can support only onsided defi for pair. 

Direction is kind of more serde friendly. For

Afaik Hydra have buy/sell in API traits, but stores all these in single collection with direction.  

Solana Serum (whole ecosystem is binary oriented and less abstact) has order(direction).

So it seems buy/sell is accepted in Polkadot as interface.

Reply:
Currently I think `buy` can not be implemented. It requires `CurveAmm` trait to expose `buy`, `sell` like interface. And possibly `get_dx` like function in pallets. if we want `buy` `sell` then perhaps its batter to have separate PR enhancing AMMs for above functions.

Reply:
I think with adding `get_dy` in `CurveAmm` should enable writing `buy`, `sell`. 
But I wanted to know why we need `buy`, `sell` instead exchange? To me it seems that UI can make RPC call to pallet invoking `get_dy()` function and get amount of token that can be received for given input amount. And flipping the direction just requires to change in `i` and `j` in `get_dy()` Hence exchange should be sufficient.

Reply:
afaik limit os sell is this price or larger. limit of buy is this price or smaller. 
can rename into from to?
user may not care what swap he uses, stable or uniswap, so pallets seems should have api for pairs regardless of amm, for pairs.

Reply:
Sorry that you got confused, but  `CurveAmm` is name for generic trait which has `exchange` function. It is not specific to Curve.Finance . We may consider to rename it in separate PR.
`exchange` function does not require price to be passed. It just needs balance i,e how many token X you want to swap, based on that amount you get token Y.

Reply:
@KaiserKarel @dzmitry-lahoda for buy/sell I have separate PR which adds functionality to CurveAmm trait first. Progess can be tracked at https://github.com/ComposableFi/composable/pull/550